File: gunkim/leetcode/tree/124.Binary Tree Maximum Path Sum.java
public class Main {
    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println(solution.maxPathSum(new TreeNode(1, new TreeNode(2), new TreeNode(3))));
    }
}

class Solution {
    public int max = Integer.MIN_VALUE;

    public int maxPathSum(TreeNode root) {
        getMax(root);
        return max;
    }

    private int getMax(TreeNode root) {
        if (root == null)
            return 0;
        int left = Math.max(getMax(root.left), 0);
        int right = Math.max(getMax(root.right), 0);
        if ((left + right + root.val) > max) {
            max = (left + right + root.val);
        }
        return Math.max(left, right) + root.val;
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

File: anb/leetcode/tree/124. Binary Tree Maximum Path Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int ans;
    int dfs(TreeNode* root) {
        ans = max(ans, root->val);
        if(root->left == nullptr && root->right == nullptr)
            return root->val;
        
        int ret, left = 0, right = 0;
        if(root->left)
            left = max(0, dfs(root->left));
        if(root->right)
            right = max(0, dfs(root->right));
        ret = root->val + max(left, right);
        ans = max({ans, ret, root->val + left + right});
        return ret;
    }
public:
    int maxPathSum(TreeNode* root) {
        ans = -1e9;
        dfs(root);
        return ans;
    }
};